bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -280,16 +280,21 @@ void ReplicaImp::onMessage<ReplicaAsksToLeaveViewMsg>(ReplicaAsksToLeaveViewMsg
 }
 
 bool ReplicaImp::checkSendPrePrepareMsgPrerequisites() {
+  if (!isCurrentPrimary()) {
+    LOG_WARN(GL, "Calling checkSendPrePrepareMsgPrerequisites in non primary replica!");
```

Comment:
Let's make this an assert. I think the statement "a non-primary will never send a PrePrepare" is a system invariant.